File: app/src/main/java/co/early/password123/ui/common/anim/Cordic.java
package co.early.password123.ui.common.anim;

import android.animation.TimeInterpolator;


public abstract class Cordic implements TimeInterpolator {

    // table data normalized to 1000
    private static int[] cordicTable;

    @Override
    public float getInterpolation(float input) {

        if (cordicTable == null){
            cordicTable = getNormalised1000CordicTable();
        }
        return 1f - (float) cordicTable[(int) (input * 1000)]/1000;
    }

    abstract int[] getNormalised1000CordicTable();

}

File: app/src/main/java/co/early/password123/ui/common/anim/Cordic.kt
package co.early.password123.ui.common.anim

import android.animation.TimeInterpolator


abstract class Cordic : TimeInterpolator {

    internal abstract val normalised1000CordicTable: IntArray

    override fun getInterpolation(input: Float): Float {

        if (cordicTable == null) {
            cordicTable = normalised1000CordicTable
        }
        return 1f - cordicTable!![(input * 1000).toInt()].toFloat() / 1000
    }

    companion object {

        // table data normalized to 1000
        private var cordicTable: IntArray? = null
    }

}
